docs: update hydration html validation information (#56343) (#56344)
angular/angular: adev/src/content/guide/hydration.md

### Valid HTML structure

> There are a few cases where if you have a component template that does not have valid HTML structure, this could result in a DOM mismatch error during hydration. This mismatch happens because browsers when parsing invalid HTML, will produce a valid DOM by making some change on the way. For example, by adding end tags when non-nestable elements are found. This introduces changes between an SSR generated HTML and the final DOM representation..

Reviewer: ```suggestion
There are a few cases where if you have a component template that does not have valid HTML structure, this could result in a DOM mismatch error during hydration. This mismatch happens because when parsing invalid HTML, browsers will recover from invalid DOM by changing it to something that is valid. For example, browsers will add end tags when non-nestable elements are found. This introduces changes between an SSR generated HTML and the final DOM representation..
```

Reviewer: I think this change would make this section much clearer.

> There are a few cases where if you have a component template that does not have valid HTML structure, this could result in a DOM mismatch error during hydration. This mismatch happens because browsers will edit invalid HTML in an attempt to make it usable, for example, by adding in new element end tags when non-nestable elements are found. This introduces changes between an SSR generated page and the HTML in the elements tab on Chrome Dev Tools.

Reviewer 2: You mixing the concept of HTML (the file that is loaded, and the DOM the internal representation which what you see in the devtools,)

```suggestion
There are a few cases where if you have a component template that does not have valid HTML structure, this could result in a DOM mismatch error during hydration. This mismatch happens because browsers when parsing invalid HTML, will produce a valid DOM by making some change on the way. For example, by adding end tags when non-nestable elements are found. This introduces changes between an SSR generated HTML and the final DOM representation..
```

* `<a>` inside an `<h1>`
* `<a>` inside another `<a>`

> If you are uncertain about whether your HTML is valid, you can use a [syntax validator](https://validator.w3.org/) to check it. Additionally, at build time HTML validation tools like the [html-validate](https://www.npmjs.com/package/html-validate) NPM package with the element-permitted-parent rule enabled can help to automate the validation process. Note that using content projection can circumvent template level HTML validation.

Reviewer 2: I'm not sure this is a great recommendation as with the new control flow, templates are not valid HTML anymore.

Reviewer 3: @JeanMeche It does seem to work with both new and old template syntax. e.g.
 `@if (user.isHuman) {
    <human-profile [data]="user" />
} @else if (user.isRobot) {
    <!-- robot users are rare, so load their profiles lazily -->
    @defer {
        <robot-profile [data]="user" />
    <a href="#">
        {{ callText }}
        <a href="#">
            {{ callText }}
        </a>

    </a>

}
} @else {
    <p>The profile is unknown!
        }`

throws this error:
  27:10  error  <a> element is not permitted as a descendant of <a>  element-permitted-content

Reviewer 3: (FWIW the line number is correct from the input I gave the validator too)

Author: @JeanMeche I can remove the section on how to fix it at build time?

Reviewer: Like @JeanMeche said, there's a lot of things about Angular templates and what Angular generates that would fail in a validator, but wouldn't cause hydration mismatches. So this recommendation could cause some people confusion, even though it does help resolve some cases, like your example.

Author: @thePunderWoman I agree that it is confusing, validators aren't aware of ng-container inside a <ul> is an example that we came across a few times. Our site has hundreds of different page components with highly dynamic content, using a template validator has saved us lots of time trying to enable hydration. 

As an aside, If we had a way to enable hydration at a route level (I don't think there is?), we could enable hydration page by page - at the moment if we disable hydration at the page component level if there is invalid html in any component the page will crash anyway because hydration is enabled in the root component.

Should I just remove those changes completely? Thanks.

Reviewer 2: I would remove that part yes.

Reviewer 2: Are you still looking at land this change ?
